### ISSUE TITLE
Even Faster Edge Candidate Search for Loki and Meili

### ISSUE THREAD

**Issue by kevinkreiser** [valhalla/valhalla#2529]
Long ago in a land far away, we wrote the code which "bins" all the edges in the graph into a grid that is finer than the smallest tile that we support. These bins are .05x.05 degrees in size which means at the equator (ie worst case) the diagonal covers 7880m in distance.

Because of this coarseness, meili, the map matching component, creates a secondary cache at 100mx100m grid on the fly and must deal with purging this etc. It does this because even though the algorithm in loki is fast for looking up coordinates, it is not fast enough for mapmatching with large number of points. We've done experiments on letting loki do this work for meili and the performance difference were roughly an order of magnitude (after the secondary cache was warmed).

This is annoying because why would we want to maintain another cache and another code path for doing the same thing, right?

Recently I was talking to @merkispavel about how the bins work, what they do and how they are used by loki/meili. We were specifically talking about how the sorting of the bins affects the performance of the algorithms that iterate over them (https://github.com/valhalla/valhalla/pull/2515#discussion_r466685873). As I was explaining this to him i mentioned that in addition to the sorting proposed here, potentially other sorting could make sense, specifically if we sort the bins in a spatial way at a smaller resolution. this isnt really possible because there is no way to tell the boundaries (where one partition ends and another begins) in the sorted list of IDs but a different idea hit me..

The GraphIDs have 18bits spare. Previously I had considered storing reachability information in these bits but precomputing that is rather difficult with the ability to change the accessabilty of a given mode of transport on the fly.. So what if we used the extra bits instead as a mask to tell what subregions of the bin a given edge crosses?

Since we have 18bits spare, we could further chop the .05x.05 degree bin into 16 smaller pieces (4x4). Each sell in this sub-bin would have a bit that we could flip in the mask on the extra 16 bits of every edges edge id. The algorithm would work similar to the way it does now, but instead of bins we would want to iterate over sub bins in closest first order. Technically this change would be backwards compatible because the old code wouldnt care about the spare bits being used and we could make new code fall back to the old algorithm when it detects that the sub bins arent set (ie the spare bits are set to 0).

Doing this would reduce the maximum diagonal from 7880m down to 1970m at the equator (worst case). Its still not 100m, but 4x less area covered. This should speed up loki::Search and hopefully make it fast enough to use in place of the secondary finer cache we have in meili::candidate_search

**Comment by dnesbitt61** [comment]
this is storing extra information in the GraphIds stored within bins, not storing this information in directed edges or nodes?

**Comment by kevinkreiser** [comment]
@dnesbitt61 exactly correct. we could even make a new struct called spatialgraphid or some such thing that claims those bits just for the id's stored in the bins and nowhere else

**Comment by kevinkreiser** [comment]
## revisiting this idea

ive been mulling this over for a long time now and have never written back the evolution of my thoughts... what i said earlier:

> this isnt really possible because there is no way to tell the boundaries (where one partition ends and another begins) in the sorted list of IDs

this isnt exactly true. its true that if we just did a sub binning and nothing else, we wouldnt know where the boundaries are. however, we could store the boundary information, that is the index of the first entry in each sub bin as we do for the existing bins. to make it match what map matching uses though would be cost prohibitive:

    5566m / 56 = 99m
    25 bins * (56x56 sub bins) = 78400 sub bins
    4 bytes per sub bin index * 78400 sub bins * 207k tiles = 64gb
    86gb planet tar + 64gb hires index = 150gb/86gb = 75% increase in data size

the benefits are its a no parse direct indexing. because its part of the tileset, it will be memory mapped and shared by all processes/threads that access the same map (grid cache is not shared so really balloons memory plus has the burden of doing allocs and deallocs when cache eviction fires). but in practical terms this is too much to pay.

the idea above with the mask isnt a bad one but it leaves us with a fairly coarse coverage. its basically saying we can divide our 5x5 bins into (5*4)x(5*4) bins. doing this with a mask is needlessly complicated, you could store the secondary binnings indices for only 1.6k extra per tile which is an extra 300mb for the whole tileset, ie less than 1% increase in tile size. so in my opinion we have 2 options.

## find a compromise between tile size increase and granularity of the index

// TODO: give my opinion on the optimal compromise

## use the spare bits to store a very fine grained index and spend cpu to traverse it

// TODO: explain what this would look like

**Comment by chrstnbwnkl** [comment]
I just ran a quick scan of a planet extract using a grid size of 0.005 degrees. That would give a length of about 556 meters per tile at the equator: 
```
earth's circumference in meters / 360 / (1/0.005) = 556.6
```

On average, each edge shape intersects with 1.42 tiles:

```
total edge count: 1012784848
intersected tile count: 1439193062
```

**Comment by kevinkreiser** [comment]
lets combine some of the ideas above:

1. we have 18bits, what if we used that to store a spatial primitive for fast rejection
1. lets go with the circle since it has only 3 components and a higher likelihood of fitting in the space
1. because it goes in the space we already have its free spacewise
1. because old data has this zero'd, our new algorithm can treat that as a sentinel that says you have to look at the shape, keeping forward compatibility
1. what the heck do we put in the bits:
   1. we need some number of bits to signify a radius to figure out how many we need we need to get some statistics 
      1. need to loop over all the edges in the data and compute minimum bounding circles, so a ll and a radius per edge
      2. look at this histogram and determine the cutoff (ie max radius at which we get most of the edges in the data covered)
      3. determine the resolution for the radius, ie each increment of that value is worth 10m or 20m, keeping in mind we need the rest of the bits to represent the centroid
   1. we need some number of bits to signify the centroid of the circle. because we have so few bits we need to record this centroid as an offset from some other easily gotten reference point. the origin of the map is over too large a range so we cant use that. the origin of the tile wont work because the centroid of the circle could be outside the tile and its also too coarse. the origin of the bin also wont work because again the centroid could be outside the bin and again too coarse. what we do know for sure is that within the minimum bounding circle the begin node of the edge will be in the tile (except tweeners which will get a 0'd out sentinel value) which means we can easily get the nodes lat lon and so we can reprsent the circles centroid as an offset from that lat lon. and we know that it will be less than the largest radius that we support (mentioned above)

so to figure this out we really just need to run over the data, computing the min circle for each to get the histogram of radii. once we have that we can tell what we want our max supported radius to be. this will dictate the number of bits we'll have both for the radius and the offset and we'll be able to see what kind of resolution for each we can support. we can vary this resolution to make sure to fit the data into these 18 bits.

**Comment by kevinkreiser** [comment]
and sadly since the offset from the node to the centroid of the circle can be in any direction we will need to use a bit for sign to make the number a signed one

**Comment by dnesbitt61** [comment]
Yes, that will impact precision for sure. I think there will be some radius value and offset lat,lng values that can be supported while retaining adequate precision for edges below this maximum. Hopefully a large percentage of edges are below the maximum and can have somewhat precise center and radius. Those that exceed some maximum radius (and this center?) may have to be skipped (meaning they will have their shape evaluated) or have some corollary record at the end of the tile that indicates a more precise center and radius for the "really long" edges. I believe bounding radius is guaranteed to be less than half the edge, so edge lengths can be a reasonable proxy for bounding circle radius. I have included a histogram of edge counts within 100m bins. 

[EdgeLengths.txt](https://github.com/user-attachments/files/17667097/EdgeLengths.txt)

**Comment by dnesbitt61** [comment]
Not sure why, in your discussion above, we cannot store the offset from the center of the bin? I don't see that the offsets to the center of the bounding circle would be much different in range from an end node of the edge rather than from the center of the bin. An edge stored in different bins will have different offsets stored within the GraphIds. This is a bit more work during data generation, but I think it will be much less work during search. If I have this correct, the search method will:

- Search finds a list of bins near the location
- For each bin, iterate over the GraphIds of edges within the bin. For each:
- Find the center using offsets (from the edge's extended GraphId in the bin) from the bin center.
- Find the squared distance to the search location.
- Reject the edge if the bounding circle radius (from extended GraphId in the bin) plus some error threshold (based on precision of offsets and radius) is less than the squared distance to the location.

Hopefully this rejects a large proportion of candidate edges.
  
This approach avoids having to access directed edges and especially nodes within the tile (and possible other tiles) except for those not rejected. Finding the start node of a directed edge is not the most efficient operation if I recall.

**Comment by kevinkreiser** [comment]
the reason i didnt consider this option was because its too coarse. a bin is 5566m across, im assuming that bounding circles will in the median be have much smaller radii. since the begin node will be within the circle then the offset will be no larger than the max radius. this lets us more efficiently use the spare bits, but it does mean a look up on the node rather than just using the bin center.

**Comment by dnesbitt61** [comment]
makes sense. I had not done any estimation of bin size and precision. But won't other tiles have to be accessed sometimes to get the start node of an edge? Directed edges only store their end node. So you have to read the opposing directed edge (sometimes in another tile) and get its end node. If we only had another 8 bits!

Storing a new record at the end of the tile with bounding volumes of directed edges also seems feasible as well. If we use a 64 bit word to encode lat,lng and radius it would add about 5GB (about 7% increase). Or perhaps a 32 bit word with offsets to the tile base could work (leading to a smaller size increase). Then it is a performance vs. memory use tradeoff. I'm not sure which I would choose for embedded apps. Probably performance - especially if faster than meili bin indexing (getting rid of that would offset some of the memory increase).

**Comment by kevinkreiser** [comment]
ah fair point! we can store the edge id of the edge whose begin node is guaranteed to be in the tile but getting the begin node might mean leaving the tile, crap :smile:

**Comment by kevinkreiser** [comment]
Maybe we could store the edge I'd of the edge whose end node is in the tile and for those who aren't since you'd have to get the shape anyway you could just use the first point in the shape. certainly no slower than today for the small number oh edges that cross out of the tile. I guess that would work for the tweeners too! Would be nice to get stats on how much of that there is

**Comment by dnesbitt61** [comment]
That seems like a good idea! I was thinking tile->GetBin would return a small(er) candidate set of edges, rejecting many/most based on the bounding circle information for any edges whose end node is within the current tile. Those whose end node is outside, could be returned (without evaluating bounding circle) and these would go through the more complex costing/filtering/shape processing. I guess bounding circle checks could be performed on these as well to potentially avoid processing shape.

**Comment by kevinkreiser** [comment]
the way loki currently works, and optimization performed a long time ago, is that we iterate the bins in the outer loop and the locations in the inner loop, so we cant really do it in the bin getting part (without changing how stuff works) because we need all the stuff in the bin to be checked against all the locations who are in play for that bin. the navitia people did this optimization and the gist is that it allows us to open the shape up for a given edge only once per call to loki::search rather than possibly multiple times each time we iterate over a different location (when the locations are in the outer loop). because of this i think we have to keep returning all the ids in the bin and then loki will have to do the circle rejection checks, which i think is perfectly fine, loki search isnt all that large or complex at the moment

**Comment by dnesbitt61** [comment]
I see. Yet another possibility. Use the original bin subdivision idea (4x4 sub-bin Ids). In theory this should reject most (perhaps 90%?) of the candidate edges retrieved from the bins using simple mask checks.  One could also store the bounding circle of each edge within EdgeInfo (use the last spare in EdgeInfoInner to indicate has_bounding_circle). This would only be accessed after costing/filtering occurs but could reject long edges that can be costly when analyzing shape. This would add 32 bits per EdgeInfo. This would increase data size by perhaps 2% (since there are roughly half the number of EdgeInfo records as DirectedEdge records).

**Comment by kevinkreiser** [comment]
i was of the opinion that 5566/4~=1400m was way too coarse to work for map matching performance-wise. currently in map matching we use a spatial grid of 100m. this is so that for a given point we only have to consider a minimal number of edge candidates for the usually hundreds of map matching input points we have to process. i suppose we could vet this assumption by changing the grid size to 1400m and seeing the difference in map matching performance?

**Comment by chrstnbwnkl** [comment]
I've calculated minimum bounding circles for all edges in a recent planet extract, and it seems like @kevinkreiser's initial ballpark estimate of a 350 meter threshold for the radius wasn't too bad.
![acc](https://github.com/user-attachments/assets/741694e3-8bf3-4eb5-abe5-bca7224ca533)
>  Note that this uses a radius in degrees converted to meters based on the center's latitude. 

In order to choose a good threshold, we should probably look at how big the center offsets from the begin node could be at higher latitudes. Unless you care about following the footsteps of Roald Amundsen, we can probably discard anything beyond 85° of latitude? :smile: 

| share of edges (%)  | radius (m) | radius (degrees of longitude at 85° latitude) |
| ------------- | -------------- |----|
|  97% | 374  | 0.039 |
|  98% | 491  | 0.051 |
| 99%  | 750  | 0.078 | 

### Example: 374m threshold, 2 bits for radius, 2*8 bits per offset to center

For the offset, we get 7 bits plus one for the sign bit. That would give us a precision of 33.69 meters at the equator (higher in higher latitudes):
```
(0.039/128) * 110567.0 (metersPerDeg) =  33.69 
```

That leaves 2 bits or 4 values for the radius. One possibility would be to go by percentiles so we have equal sized groups. Plus we would need to pad each value with the offset resolution. For 374 meters that would be approximately:

| percentile  | radius (m) | radius with offset (m) |
| ------------- | -------------- |-----|
|  25% | 11 | 44.69 |
|  50% | 28 | 61.69 |
| 75%  | 67 | 100.69 |
| 100%  | 374 | 407.69 |

**Comment by kevinkreiser** [comment]
> we should probably look at how big the center offsets from the begin node could be at higher latitudes

this didnt make sense to me but i could be missing your point. heres how i understood it.

the more extreme the latitude away from the equator the closer the lines of longitude become to each other hence the smaller the number of degrees to cover the same distance (in the x axis (longitude)) when compared to the equator. ie. a longitude/x coord of 30 degrees vs a longitude/x coord of 31 degrees has a distance that is larger at the equator than up near the pole. this means that the center offsets in the x axis, to fit inside that 374m radius will be shorter/smaller number (in degrees) up near the pole than down at the equator. so yeah i think the equator is really where it matters, how many degrees that max radius covers at the equator should determine our resolution for offset.

though perhaps your point is less about the max value and more about resolution. that is, if up near the poles 374m covers so little distance in degrees units then our offset will always be too coarse for those latitudes if we base them off of the equator where we have plenty of degrees to divvy up? is that what you meant by that?

**Comment by chrstnbwnkl** [comment]
Maybe this is a sign that I've had enough screen time for today but isn't it the other way around? :smile: Simply put, if you take a yard stick, travel to the north pole and put it down somewhere dangerously close to the geographical pole (along a line of latitude), it will take up several degrees of longitude right? Doesn't this then mean that a radius of 374 meters converts to a higher value in degrees further up at the poles than down at the equator and hence a larger maximum offset we'd have to support?

**Comment by kevinkreiser** [comment]
@chrstnbwnkl yep youre right! i shouldnt do two drastically different things at once it seems :smile:

**Comment by chrstnbwnkl** [comment]
Given that, we can also mess around with the cutoff latitude. I didn't check that yet, but I'd assume there are very few settlements even beyond 75° on both hemispheres (sorry Svalbard).

**Comment by kevinkreiser** [comment]
if you wrote out the centroid latitudes when generating them you could do it empirically to gaurantee a reasonable max coverage 97% or something

**Comment by chrstnbwnkl** [comment]
Luckily I did! Even a latitude threshold of 70° would include 99.99% of all edges. Beyond that, there don't seem to be any large human settlements anyway. According to Wikipedia, the largest settlement beyond 70° lat is [a little town called Hammerfest, Norway)[https://en.wikipedia.org/wiki/Hammerfest_(town)].

**Comment by chrstnbwnkl** [comment]
Another point that just crossed my mind w/r/t minimum bounding circles: seems like 60% of all edges in my planet extract have a shape size of 3 or 2. For these, we might as well just look at the shape instead right? So maybe the 4x4 bin mask is the better option after all?

**Comment by kevinkreiser** [comment]
wow 60%!? that is wild. what does that histogram look like?

**Comment by kevinkreiser** [comment]
planet extract? is it the full planet? my gut doesnt agree with this conclusion:

> 60% of all edges in my planet extract have a shape size of 3 or 2. For these, we might as well just look at the shape instead right?

getting the shape requires us to go to a different location in memory which means paging cache in from either disk or ram to cpu registers. when we evaluate the bins we are already looking at the exact 64bit value that would have the bounding circle in it so we dont have to go anywhere in the vast vast majority of cases (non tweeners and edges leaving tiles, should get stats on this but i bet its like 4% of bin entries). also the space is currently unused so its like free to use it, the best price.

the mask is still very coarse as mentioned above, we'll still be going to shape checks a decent amount.

better than conjecture though, perhaps it would be worth implementing these and benchmarking them? its not that much work right? we already have loki_benchmark so we just need code to modify the bins for both approaches and code to use the modified bins for both approaches.

how has a strong opinion on next steps? my personal opinion is we have a lot of good ideas, we should just straight up make branches of each and try them out :smile:

**Comment by kevinkreiser** [comment]
ah crap i just realized that its not just tweeners and edges that leave the tile (if we base the offset off the end node) its all edges that arent level 2 which is a massive amount of the edges.. more thought needed if we want to fast reject the edges right away off data attached to the bin graphid record.

**Comment by kevinkreiser** [comment]
i think it could still work using the bin centroid as the point from which we offset up to some max offset (4000m would go to the bin corner at the equator). we can brainstorm some more about what this strategy would exactly look like. mask strategy is straightforward i think so could write that code right away for benching

**Comment by chrstnbwnkl** [comment]
> getting the shape requires us to go to a different location in memory which means paging cache in from either disk or ram to cpu registers. 

Right, I forgot about edge info being at an entirely different location in the tile :smile:

**Comment by kevinkreiser** [comment]
ok so we talked this morning and we're thinking an experiment with the circles approach like so:

- each bin has a hypotenuse of about 4km
- we know empirically that we get most edges within a circle of 375m
- so we can bound the centroid of a bin by about 4375m
- which means we can make offset increments of about 4375/128 (7bits) = 34m
- for the radius increments we should still use the nonlinear increments with the 2bits
- first increment should be maybe about 15 or so so that it covers the whole inaccuracy of the offset (ie a sub bin really)
- we'll want to compute the lat scaling for each bin we process, this is an extra cosine per bin
- even for large input we dont process that many bins typically so this should be noise but it might play a perf role
- but anyway this lets us work in meters essentially dependent on the actually latitude where we are looking

we should implement and bench this

**Comment by dnesbitt61** [comment]
is there a branch for this? In particular, code to generate a bounding circle for a lat,lng polyline?

**Comment by kevinkreiser** [comment]
goes has a method built in and @chrstnbwnkl is using that. not sure if he pushed up the code though

**Comment by chrstnbwnkl** [comment]
Yeah I used GEOS to quickly hack this together, script is here: https://github.com/chrstnbwnkl/valhalla/blob/cb-bin-stats/src/mjolnir/valhalla_min_bounding_circles.cc

**Comment by dnesbitt61** [comment]
Looking through uses of Loki Search I came across this in trace_route_action.cc:

/```
/ TODO: remove this, it was a hack to support display_ll for map matching, what we can do is
// actually use the display_ll now as its  used in loki::Search
void loki_worker_t::locations_from_shape(Api& request)
```
does this mean that this method is not needed? I commented out on our fork and didn't see any differences (but we customize our route response).

**Comment by dnesbitt61** [comment]
It seems that constructing pps is perhaps a bottleneck when mmap is not used. See comparison below.
When running the server (only 2 locations) with PA data (no mmap):
2024/11/14 15:14:29.074486 [INFO] Set pps locations took: 5 ms
2024/11/14 15:14:29.074633 [INFO] handler search: pps size = 2 front has bin: 1
2024/11/14 15:14:29.075081 [INFO] Search time= 5 ms
When running with mmap planet (no bounding circles):
2024/11/14 15:15:48.798071 [INFO] Set pps locations took: 0 ms
2024/11/14 15:15:48.798107 [INFO] handler search: pps size = 2 front has bin: 1
2024/11/14 15:15:48.808416 [INFO] Search time= 10 ms
Running valhalla_benchmark_loki:
PA - no mmap:
2024/11/14 15:19:01.086322 [INFO] Set pps locations took: 11421 ms
2024/11/14 15:19:01.088762 [INFO] handler search: pps size = 8714 front has bin: 0
2024/11/14 15:19:01.088772 [INFO] Search time= 11424 ms
planet mmap:
2024/11/14 15:20:07.232195 [INFO] Set pps locations took: 1660 ms
2024/11/14 15:20:07.234386 [INFO] handler search: pps size = 8714 front has bin: 0
2024/11/14 15:20:07.234394 [INFO] Search time= 1662 ms

When using files rather than mmap, it seems each projector_wrapper constructor (in the emplace call) is taking 1-2ms (added timing). Any thoughts why this differs between mmap and file data?

**Comment by kevinkreiser** [comment]
is this measuring the first load by chance? both mmap and connectivity map are lazy loaded, that means the first load is very very slow as it scans the file or directory (iirc). in practice this is amortized away as subsequent requests do not pay this penalty.

regarding that hack yeah we could make use of display_ll if the input for the request was locations instead of polyline encoded shape

**Comment by dnesbitt61** [comment]
sorry, I had a problem with my branch. Ran again on master and while using tiles is slightly slower than mmap it isn't as dramatic as above. Sorry about the noise!
Thanks for the info on locations_from_shape - I don't think we need it for our use case.

**Comment by kevinkreiser** [comment]
no problem! thank you for helping us get a handle on the perf characteristics and options for one of my favorite avenues of potential speed up :smile:

**Comment by dnesbitt61** [comment]
Now that valhalla_benchmark_loki finally works. Using 8760 vertices from a GPX file in PA:
master (using planet with individual files):
Succeeded searches on cached tiles: Median: 0.058165ms

On a branch with us data with lat,lng,radius added to EdgeInfo. Code added to loki/search.cc to skip edges where location is outside the bounding circle of the edge.
Succeeded searches on cached tiles: Median: 0.007541ms

This seems to be 6x faster if my math is right.

**Comment by dnesbitt61** [comment]
Also, when generating US data the bounding circle code crashed when building tiles (probably due to the way I added it - likely not in a thread safe manner). Anyway, I used a variation of Ritter's method to create the bounds. This isn't guaranteed to be the minimum bounding circle but is supposed to be close. When I tested routes using the Leaflet demo app I saw that most times only a handful of edges actually had shape decoded and searched. If the locations were in rural areas, often only 1 edge was evaluated!

**Comment by kevinkreiser** [comment]
nice! 7.5x, if i remember right, is just shy (roughly 10x) of what we get on the map matching endpoint if you hit it first to fill its "grid cache" and then hit it again to use the cache and then hack in loki search to handle that part and do the same thing. imho 7.5x sounds close enough!

**Comment by dnesbitt61** [comment]
I pushed up a branch, bv_edgeinfo, with code to populate bounding circles in EdgeInfo. I added a new builder EdgeBoundsBuilder stage which is done after building elevation. This was the easiest way to get bounds on all edges - including shortcuts, transit, etc. I am using 3 floats which adds about 6.7% size increase. We could compress this down to 64bits which would be just over 4% size increase. I tested that the server code works (at least for routes using the leaflet test app) with both new data and with old data (w/o bounds).
In search.cc I am using 50m threshold above the bounding radius. I think we should use location.search_cutoff - this I think defaults to 50m for map matching, but 35000 for other cases. We may want to add this as a run time arg to valhalla_benchmark_loki and pass this in on the locations.
